### ISSUE TITLE
Failed to install `pg` for local development

### ISSUE THREAD

**Issue by DaveParr** [thepracticaldev/dev.to#8267]
I'm trying to follow the [instructions here](https://docs.dev.to/installation/mac/) on getting setup locally.

I've setup postgres from [postgres.app](https://postgresapp.com/), and I think I've also set up the config [correctly as here](https://docs.dev.to/installation/postgresql/) after using [this approach](https://www.codementor.io/@engineerapart/getting-started-with-postgresql-on-mac-osx-are8jcopb) to create a user. 

I think the problem stems from further up the install chain though, as I am getting this response from `bundle install`:

```sh
bundle install 
Fetching gem metadata from https://rubygems.org/......
Using rake 13.0.1
Using concurrent-ruby 1.1.6
Using i18n 1.8.2
Using minitest 5.14.1
Using thread_safe 0.3.6
Using tzinfo 1.2.7
Using activesupport 5.2.4.3
Using builder 3.2.4
Using erubi 1.9.0
Using mini_portile2 2.4.0
Using nokogiri 1.10.9
Using rails-dom-testing 2.0.3
Using crass 1.0.6
Using loofah 2.5.0
Using rails-html-sanitizer 1.3.0
Using actionview 5.2.4.3
Using rack 2.2.2
Using rack-test 1.1.0
Using actionpack 5.2.4.3
Using nio4r 2.5.2
Using websocket-extensions 0.1.4
Using websocket-driver 0.7.1
Using actioncable 5.2.4.3
Using globalid 0.4.2
Using activejob 5.2.4.3
Using mini_mime 1.0.2
Using mail 2.7.1
Using actionmailer 5.2.4.3
Using activemodel 5.2.4.3
Using arel 9.0.0
Using activerecord 5.2.4.3
Using active_record_union 1.3.0
Using activerecord-import 1.0.5
Using mimemagic 0.3.5
Using marcel 0.3.3
Using activestorage 5.2.4.3
Using acts-as-taggable-on 6.5.0
Using acts_as_follower 0.2.1 from https://github.com/thepracticaldev/acts_as_follower.git (at /Users/davidparr/Documents/GitHub/dev.to/vendor/cache/acts_as_follower-288690cd99bc@288690c)
Using public_suffix 4.0.5
Using addressable 2.7.0
Using execjs 2.7.0
Using autoprefixer-rails 9.7.6
Using method_source 1.0.0
Using thor 0.20.3
Using railties 5.2.4.3
Using momentjs-rails 2.20.1
Using datetime_picker_rails 0.0.7
Using jquery-rails 4.4.0
Using kaminari-core 1.2.1
Using kaminari-actionview 1.2.1
Using kaminari-activerecord 1.2.1
Using kaminari 1.2.1
Using ffi 1.12.2
Using sassc 2.2.1
Using sprockets 4.0.0
Using sprockets-rails 3.2.1
Using tilt 2.0.10
Using sassc-rails 2.1.2
Using selectize-rails 0.12.6
Using administrate 0.13.0
Using after_commit_action 1.1.0
Using errbase 0.2.0
Using safely_block 0.3.0
Using ahoy_email 1.1.0
Using mime-types-data 3.2020.0512
Using mime-types 3.3.1
Using multi_xml 0.6.0
Using httparty 0.18.0
Using json 2.3.0
Using algorithmia 1.1.0
Using amazing_print 1.1.0
Using ancestry 3.0.7
Using approvals 0.0.24
Using ast 2.4.0
Using aws-eventstream 1.1.0
Using aws-partitions 1.322.0
Using aws-sigv4 1.1.4
Using jmespath 1.4.0
Using aws-sdk-core 3.97.0
Using aws-sdk-lambda 1.42.0
Using aws_cf_signer 0.1.3
Using descendants_tracker 0.0.4
Using ice_nine 0.11.2
Using axiom-types 0.1.1
Using bcrypt 3.1.13
Using benchmark-ips 2.7.2
Using coderay 1.1.2
Using better_errors 2.7.1
Using html_tokenizer 0.0.7
Using parser 2.7.1.3
Using smart_properties 1.15.0
Using better_html 1.0.15
Using bindex 0.8.1
Using debug_inspector 0.0.3
Using binding_of_caller 0.8.0
Using chartkick 3.3.1
Using blazer 2.2.4
Using msgpack 1.3.3
Using bootsnap 1.4.6
Using brakeman 4.8.2
Using browser 4.0.0
Using environs 1.1.0
Using multipart-post 2.1.1
Using faraday 1.0.1
Using faraday_middleware 1.0.0
Using hashie 4.1.0
Using multi_json 1.14.1
Using yajl-ruby 1.4.1
Using buffer 0.1.3
Using buftok 0.2.0
Using uniform_notifier 1.13.0
Using bullet 6.1.0
Using bundler 2.1.4
Using bundler-audit 0.6.1
Using byebug 11.1.1
Using regexp_parser 1.7.0
Using xpath 3.2.0
Using capybara 3.32.2
Using mini_magick 4.10.1
Using ruby-vips 2.0.17
Using image_processing 1.10.3
Using carrierwave 2.1.0
Using fastimage 2.1.1
Using carrierwave-bombshelter 0.2.2
Using caze 0.2.2
Using childprocess 3.0.0
Using cld 0.8.0
Using http-accept 1.7.0
Using unf_ext 0.0.7.7
Using unf 0.1.4
Using domain_name 0.5.20190701
Using http-cookie 1.0.3
Using netrc 0.11.0
Using rest-client 2.1.0
Using cloudinary 1.14.0
Using coercible 1.0.0
Using coffee-script-source 1.12.2
Using coffee-script 2.4.1
Using coffee-rails 4.2.2
Using connection_pool 2.2.2
Using counter_culture 2.5.1
Using safe_yaml 1.0.5
Using crack 0.4.3
Using dante 0.2.0
Using ddtrace 0.36.0
Using get_process_mem 0.2.5
Using heapy 0.1.4
Using memory_profiler 0.9.14
Using mini_histogram 0.1.3
Using ruby-statistics 2.1.2
Using enumerable-statistics 2.0.1
Using unicode_plot 0.0.4
Using derailed_benchmarks 1.7.0
Using orm_adapter 0.5.0
Using responders 3.0.0
Using warden 1.2.8
Using devise 4.7.1
Using diff-lcs 1.3
Using distribution 0.7.3
Using docile 1.3.2
Using dogstatsd-ruby 4.8.1
Using doorkeeper 5.4.0
Using dry-core 0.4.9
Using dry-equalizer 0.3.0
Using dry-configurable 0.11.1
Using dry-container 0.7.2
Using dry-inflector 0.2.0
Using dry-logic 1.0.6
Using dry-types 1.3.1
Using dry-struct 1.3.0
Using elasticsearch-api 7.7.0
Using elasticsearch-transport 7.7.0
Using elasticsearch 7.7.0
Using eventmachine 1.2.5
Using http_parser.rb 0.6.0
Using em-websocket 0.5.1
Using email_validator 2.0.1
Using emoji_regex 3.0.0
Using envied 0.9.3
Using equalizer 0.0.11
Using rainbow 3.0.0
Using parallel 1.19.1
Using rexml 3.2.4
Using rubocop-ast 0.0.3
Using ruby-progressbar 1.10.1
Using unicode-display_width 1.7.0
Using rubocop 0.84.0
Using erb_lint 0.0.33
Using excon 0.73.0
Using exifr 1.3.6
Using factory_bot 5.2.0
Using factory_bot_rails 5.2.0
Using faker 2.11.0
Using faraday-http-cache 2.2.0
Using fast_jsonapi 1.5
Using fastly 2.5.2
Using sax-machine 1.3.2
Using feedjira 3.1.1
Using ffi-compiler 1.0.1
Using field_test 0.3.2
Using figaro 1.2.0
Using fix-db-schema-conflicts 3.0.3
Using flipper 0.17.2
Using flipper-active_record 0.17.2
Using rack-protection 2.0.8.1
Using flipper-ui 0.17.2
Using formatador 0.2.5
Using fog-core 2.2.0
Using fog-json 1.2.0
Using fog-xml 0.1.3
Using ipaddress 0.8.3
Using fog-aws 3.6.5
Using front_matter_parser 0.2.1
Using gemoji 4.0.0.rc2
Using gibbon 3.3.4
Using rb-fsevent 0.10.3
Using rb-inotify 0.10.1
Using listen 3.2.1
Using lumberjack 1.2.4
Using nenv 0.3.0
Using shellany 0.0.1
Using notiffany 0.1.3
Using pry 0.13.1
Using guard 2.16.2
Using guard-compat 1.2.1
Using guard-livereload 2.5.2
Using rspec-support 3.9.3
Using rspec-core 3.9.2
Using rspec-expectations 3.9.2
Using rspec-mocks 3.9.1
Using rspec 3.9.0
Using guard-rspec 4.7.3
Using hashdiff 1.0.1
Using honeybadger 4.6.0
Using http-form_data 2.3.0
Using http-parser 1.2.1
Using http 4.3.0
Using libhoney 1.14.4
Using honeycomb-beeline 2.0.0
Using html_truncator 0.4.2
Using htmlentities 4.3.4
Using httpclient 2.8.3
Using hypershield 0.2.0
Using inline_svg 1.7.1
Using jbuilder 2.10.0
Using sass-listen 4.0.0
Using sass 3.7.4
Using jquery-fileupload-rails 0.4.7
Using jwt 2.2.1
Using katex 0.6.0
Using launchy 2.5.0
Using libv8 7.3.492.27.1 (x86_64-darwin-19)
Using liquid 4.0.3
Using memoizable 0.4.2
Using mini_racer 0.2.14
Using nakayoshi_fork 0.0.4
Using naught 1.1.0
Using oauth 0.5.4
Using oauth2 1.4.3
Using sawyer 0.8.2
Using octokit 4.18.0
Using oj 3.10.6
Using omniauth 1.9.1
Using omniauth-oauth2 1.6.0
Using omniauth-github 1.4.0
Using omniauth-oauth 1.1.0
Using omniauth-twitter 1.4.0
Using parallel_tests 2.32.0
Using patron 0.13.3
Using percy-capybara 4.2.0
Installing pg 1.2.3 with native extensions
Errno::EACCES: Permission denied @ rb_sysopen - /Users/davidparr/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/pg-1.2.3/.gemtest
An error occurred while installing pg (1.2.3), and Bundler cannot continue.
Make sure that `gem install pg -v '1.2.3' --source 'https://rubygems.org/'` succeeds before bundling.

In Gemfile:
  pg
```

Running as `sudo` also fails on the same install, though warns me not to execute what I just did as well, but then does it anyway without checking :p

I think an issue *might* have crept in as I got a bit excited after the fork, and skipped this line in the docs:

```sh
bundle install
```

The first time and just smashed out `rails db:setup` and then `sudo gem install rails` straight after forking the repo and cloning down. Which gave me this

```sh
Fetching concurrent-ruby-1.1.6.gem
Fetching i18n-1.8.2.gem
Fetching thread_safe-0.3.6.gem
Fetching tzinfo-1.2.7.gem
Fetching zeitwerk-2.3.0.gem
Fetching activesupport-6.0.3.1.gem
Fetching rack-2.2.2.gem
Fetching rack-test-1.1.0.gem
Fetching nokogiri-1.10.9.gem
Fetching mini_portile2-2.4.0.gem
Fetching loofah-2.5.0.gem
Fetching rails-html-sanitizer-1.3.0.gem
Fetching rails-dom-testing-2.0.3.gem
Fetching builder-3.2.4.gem
Fetching erubi-1.9.0.gem
Fetching crass-1.0.6.gem
Fetching actionview-6.0.3.1.gem
Fetching actionpack-6.0.3.1.gem
Fetching activemodel-6.0.3.1.gem
Fetching activerecord-6.0.3.1.gem
Fetching globalid-0.4.2.gem
Fetching activejob-6.0.3.1.gem
Fetching mini_mime-1.0.2.gem
Fetching mail-2.7.1.gem
Fetching actionmailer-6.0.3.1.gem
Fetching marcel-0.3.3.gem
Fetching websocket-extensions-0.1.5.gem
Fetching websocket-driver-0.7.2.gem
Fetching actioncable-6.0.3.1.gem
Fetching mimemagic-0.3.5.gem
Fetching nio4r-2.5.2.gem
Fetching activestorage-6.0.3.1.gem
Fetching actionmailbox-6.0.3.1.gem
Fetching actiontext-6.0.3.1.gem
Fetching thor-1.0.1.gem
Fetching method_source-1.0.0.gem
Fetching railties-6.0.3.1.gem
Fetching sprockets-4.0.0.gem
Fetching sprockets-rails-3.2.1.gem
Fetching rails-6.0.3.1.gem
Successfully installed concurrent-ruby-1.1.6

HEADS UP! i18n 1.1 changed fallbacks to exclude default locale.
But that may break your application.

If you are upgrading your Rails application from an older version of Rails:

Please check your Rails app for 'config.i18n.fallbacks = true'.
If you're using I18n (>= 1.1.0) and Rails (< 5.2.2), this should be
'config.i18n.fallbacks = [I18n.default_locale]'.
If not, fallbacks will be broken in your app by I18n 1.1.x.

If you are starting a NEW Rails application, you can ignore this notice.

For more info see:
https://github.com/svenfuchs/i18n/releases/tag/v1.1.0

Successfully installed i18n-1.8.2
Successfully installed thread_safe-0.3.6
Successfully installed tzinfo-1.2.7
Successfully installed zeitwerk-2.3.0
Successfully installed activesupport-6.0.3.1
Successfully installed rack-2.2.2
Successfully installed rack-test-1.1.0
Successfully installed mini_portile2-2.4.0
Building native extensions. This could take a while...
Successfully installed nokogiri-1.10.9
Successfully installed crass-1.0.6
Successfully installed loofah-2.5.0
Successfully installed rails-html-sanitizer-1.3.0
Successfully installed rails-dom-testing-2.0.3
Successfully installed builder-3.2.4
Successfully installed erubi-1.9.0
Successfully installed actionview-6.0.3.1
Successfully installed actionpack-6.0.3.1
Successfully installed activemodel-6.0.3.1
Successfully installed activerecord-6.0.3.1
Successfully installed globalid-0.4.2
Successfully installed activejob-6.0.3.1
Successfully installed mini_mime-1.0.2
Successfully installed mail-2.7.1
Successfully installed actionmailer-6.0.3.1
Building native extensions. This could take a while...
Successfully installed nio4r-2.5.2
Successfully installed websocket-extensions-0.1.5
Building native extensions. This could take a while...
Successfully installed websocket-driver-0.7.2
Successfully installed actioncable-6.0.3.1
Successfully installed mimemagic-0.3.5
Successfully installed marcel-0.3.3
Successfully installed activestorage-6.0.3.1
Successfully installed actionmailbox-6.0.3.1
Successfully installed actiontext-6.0.3.1
Successfully installed thor-1.0.1
Successfully installed method_source-1.0.0
Successfully installed railties-6.0.3.1
Successfully installed sprockets-4.0.0
Successfully installed sprockets-rails-3.2.1
Successfully installed rails-6.0.3.1
Parsing documentation for concurrent-ruby-1.1.6
Installing ri documentation for concurrent-ruby-1.1.6
Parsing documentation for i18n-1.8.2
Installing ri documentation for i18n-1.8.2
Parsing documentation for thread_safe-0.3.6
Installing ri documentation for thread_safe-0.3.6
Parsing documentation for tzinfo-1.2.7
Installing ri documentation for tzinfo-1.2.7
Parsing documentation for zeitwerk-2.3.0
Installing ri documentation for zeitwerk-2.3.0
Parsing documentation for activesupport-6.0.3.1
Installing ri documentation for activesupport-6.0.3.1
Parsing documentation for rack-2.2.2
Installing ri documentation for rack-2.2.2
Parsing documentation for rack-test-1.1.0
Installing ri documentation for rack-test-1.1.0
Parsing documentation for mini_portile2-2.4.0
Installing ri documentation for mini_portile2-2.4.0
Parsing documentation for nokogiri-1.10.9
Installing ri documentation for nokogiri-1.10.9
Parsing documentation for crass-1.0.6
Installing ri documentation for crass-1.0.6
Parsing documentation for loofah-2.5.0
Installing ri documentation for loofah-2.5.0
Parsing documentation for rails-html-sanitizer-1.3.0
Installing ri documentation for rails-html-sanitizer-1.3.0
Parsing documentation for rails-dom-testing-2.0.3
Installing ri documentation for rails-dom-testing-2.0.3
Parsing documentation for builder-3.2.4
Installing ri documentation for builder-3.2.4
Parsing documentation for erubi-1.9.0
Installing ri documentation for erubi-1.9.0
Parsing documentation for actionview-6.0.3.1
Installing ri documentation for actionview-6.0.3.1
Parsing documentation for actionpack-6.0.3.1
Installing ri documentation for actionpack-6.0.3.1
Parsing documentation for activemodel-6.0.3.1
Installing ri documentation for activemodel-6.0.3.1
Parsing documentation for activerecord-6.0.3.1
Installing ri documentation for activerecord-6.0.3.1
Parsing documentation for globalid-0.4.2
Installing ri documentation for globalid-0.4.2
Parsing documentation for activejob-6.0.3.1
Installing ri documentation for activejob-6.0.3.1
Parsing documentation for mini_mime-1.0.2
Installing ri documentation for mini_mime-1.0.2
Parsing documentation for mail-2.7.1
Installing ri documentation for mail-2.7.1
Parsing documentation for actionmailer-6.0.3.1
Installing ri documentation for actionmailer-6.0.3.1
Parsing documentation for nio4r-2.5.2
Installing ri documentation for nio4r-2.5.2
Parsing documentation for websocket-extensions-0.1.5
Installing ri documentation for websocket-extensions-0.1.5
Parsing documentation for websocket-driver-0.7.2
Installing ri documentation for websocket-driver-0.7.2
Parsing documentation for actioncable-6.0.3.1
Installing ri documentation for actioncable-6.0.3.1
Parsing documentation for mimemagic-0.3.5
Installing ri documentation for mimemagic-0.3.5
Parsing documentation for marcel-0.3.3
Installing ri documentation for marcel-0.3.3
Parsing documentation for activestorage-6.0.3.1
Installing ri documentation for activestorage-6.0.3.1
Parsing documentation for actionmailbox-6.0.3.1
Installing ri documentation for actionmailbox-6.0.3.1
Parsing documentation for actiontext-6.0.3.1
Installing ri documentation for actiontext-6.0.3.1
Parsing documentation for thor-1.0.1
Installing ri documentation for thor-1.0.1
Parsing documentation for method_source-1.0.0
Installing ri documentation for method_source-1.0.0
Parsing documentation for railties-6.0.3.1
Installing ri documentation for railties-6.0.3.1
Parsing documentation for sprockets-4.0.0
Installing ri documentation for sprockets-4.0.0
Parsing documentation for sprockets-rails-3.2.1
Installing ri documentation for sprockets-rails-3.2.1
Parsing documentation for rails-6.0.3.1
Installing ri documentation for rails-6.0.3.1
Done installing documentation for concurrent-ruby, i18n, thread_safe, tzinfo, zeitwerk, activesupport, rack, rack-test, mini_portile2, nokogiri, crass, loofah, rails-html-sanitizer, rails-dom-testing, builder, erubi, actionview, actionpack, activemodel, activerecord, globalid, activejob, mini_mime, mail, actionmailer, nio4r, websocket-extensions, websocket-driver, actioncable, mimemagic, marcel, activestorage, actionmailbox, actiontext, thor, method_source, railties, sprockets, sprockets-rails, rails after 50 seconds
40 gems installed
```

The error relates to the `pg` gem, which I guess is the postgres connector, so I tried running the line both with and without the postgres local dbs running. Both had the same result.

I'm running macOS Catalina 10.15.5.

> What do I need to do to fix this pg issue, and how did I create it in the first place?

**Comment by DaveParr** [comment]
Also figured it might be a permissions on the file.folder issue, but seemingly also not resolved by `sudo chmod 777 /Users/davidparr/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/pg-1.2.3/` The folder path does appear to be present on my system though :/

**Comment by rhymes** [comment]
What happens if you follow this instruction?

> Make sure that `gem install pg -v '1.2.3' --source 'https://rubygems.org/'` succeeds before bundling.

**Comment by DaveParr** [comment]
`ERROR:  While executing gem ... (Errno::EACCES)
    Permission denied @ rb_sysopen - /Users/davidparr/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/pg-1.2.3/ext/errorcodes.def`

**Comment by rhymes** [comment]
Mmm I would try to remove the gem folder and try again. Thees are the permissions on my folders

```shell
➜  devto git:(master) ✗ ll `gem env gemdir`/gems | grep pg
drwxr-xr-x     - rhymes 15 May 10:41  pg-1.2.3
➜  devto git:(master) ✗ ll `gem env gemdir` | grep gems
drwxr-xr-x     - rhymes  4 Jun 14:37  gems
```

**Comment by DaveParr** [comment]
This is them pre-removing the gem folder, but after the chmod. Pretty similar.

```
davidparr@Davids-MacBook-Pro  ~/Documents/GitHub/dev.to   master ●  ll `gem env gemdir`/gems | grep pg
drwxrwxrwx  19 root       staff   608B  4 Jun 13:10 pg-1.2.3
 davidparr@Davids-MacBook-Pro  ~/Documents/GitHub/dev.to   master ●  ll `gem env gemdir` | grep gems
drwxr-xr-x  341 davidparr  staff    11K  4 Jun 12:43 gems
```

**Comment by DaveParr** [comment]
Sorry, to be specifc, I have 2 gems folders in this path: `/Users/davidparr/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/pg-1.2.3`

Which is the one to remove?

**Comment by rhymes** [comment]
@DaveParr for some reason you first folder belongs to root, not to you. I think that's the issue.

Run `sudo chown -R davidparr PATH_OF_THE_FOLDER`

**Comment by DaveParr** [comment]
yup, sorry! Ok, fixed that now. re-running 
gives me:

```sh
ll `gem env gemdir`/gems | grep pg 
drwxrwxrwx  19 davidparr  staff   608B  4 Jun 13:10 pg-1.2.3
```

`bundle install` then gives me:

```sh
Bundle complete! 151 Gemfile dependencies, 349 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Post-install message from ulid:

ulid gem needs to install sysrandom gem if you use Ruby 2.4 or older.
Execute `gem install sysrandom` or add `gem "sysrandom"` to Gemfile.
```

which I think means the process is complete? and seeing as this should be 2.7 that `ulid` message should able to be ignored.

I'm now trying to get the db running with `rails db:setup`, but haven't quite solved that yet. Happy to consider this closed if you want to though while I hack at it some more. Thanks for all the help!

**Comment by rhymes** [comment]
@DaveParr please use `bin/setup` as indicated by https://docs.dev.to/installation/mac/ :D

**Comment by DaveParr** [comment]
@rhymes It's running!

some other tweaks I needed to do to get it working:

* I had previously setup the [config](https://docs.dev.to/getting-started/config-env/) here and populate in with DATABASE_URL as suggested in the setup [here](https://docs.dev.to/installation/mac/) under 'Set up your environment variables/secrets' and [here](https://docs.dev.to/installation/postgresql/#configuration) I don't think this worked out well as I got an error about connecting to port 9600. However, I'm not certain right now if that was the elasticsearch, whihc I also think I needed to have running during the execution of `bin/setup`?

* I also on the first run got a HUAGE set of data being populated in the DB, but then on trying to open local host I hit an error where the browser view complained that `//= link base.js was not compiled [something something]` (sorry, I didn't write it down). It seemed to be suggesting I needed to add a line to the manifest, which I did, and voila! 🚀

Now I just need to learn basically any/all of the technologies you use. Closest I've come is interacting with Redshift dbs. Eitherway, closed ✅
